launch/launch/substitutions/python_expression.py

```diff
@@ -62,11 +66,16 @@ def expression(self) -> List[Substitution]:
         """Getter for expression."""
         return self.__expression
 
+    @property
+    def locals(self) -> dict:
+        """Getter for locals."""
+        return self.__locals
+
     def describe(self) -> Text:
         """Return a description of this substitution as a string."""
         return 'PythonExpr({})'.format(' + '.join([sub.describe() for sub in self.expression]))
```

Comment:
I think the description should be updated to reflect the modules

Reply:
Likewise, the parse up there, otherwise the XML frontend won't work

See: https://github.com/ros2/launch/blob/rolling/launch/launch/substitutions/launch_configuration.py#L68
In this case the length of data might be 1 or 2. If it's 1, be sure to set 'modules' to the default value in the parse!

Reply:
I could store the module names in order to print them as part of the Describe output, as printing `__locals` would be way too long for many modules. But then that got me thinking that we might also want to do something like this at Line 53.

> self.__locals = {}
        for x in modules:
            self.__locals[x.__name__] = vars(x)

This way, the expression locals would remain keyed for their respective modules. Then an expression would read `sys.getrefcount(str("hello world!"))` instead of `getrefcount(str("hello world!"))`, for example. This seems good in terms of clarity and avoiding name clashes, especially if someone wants to pass a custom module.

However, doing that would make this PR a breaking change, since PythonExpression currently accepts `math` definitions without the module name. That seems like a reasonable break to make as part of a new ROS2 release, but I wanted to run this idea by you all. I think it would be a good change in the longer run.

Reply:
Thinking about it even more, the new PythonExpression constructor probably ought to take the module names as a SomeSubstitutionsType, to be consistent with how substitutions generally work.

Reply:
If we do SomeSubstitutionsType we'll most likely have to take in strings as opposed to module objects directly. I am fine with this since in that case the xml frontend can be unified. I'm not too fussed about this point though.

What will have to change is the parse:
We'll need importing logic, since there isn't a way to pass modules directly in xml.

Pardon the style/naming. I'm just trying to workshop the idea
```
import importlib

    @classmethod
    def parse(cls, data: Iterable[SomeSubstitutionsType]):
        """Parse `PythonExpression` substitution."""
        if not len(data) >= 1 and not len(data) <= 2:
            raise TypeError('eval substitution expects 1 to 2 arguments')
        modules = []
        for i in data[1].split("SOME_DELIMITER_PROBABLY_COMMA"):
            module = importlib.import_module(i)
            modules.append(module)
        # the name 'modules' could change
        return cls, {'expression': data[0], 'modules': module}
```

Reply:
As for breaking, I think we can have our cake and eat it too. We can just print a warning instead and support both cases.

We can store modules as well as any local names.

More precisely, we can:
```
# Modules is probably not a good name at this point
for x in modules:
    if type(x) is module:
        if any(i for i in vars(x) in in self__locals):
            # PRINT A WARNING
        self.__locals.update(vars(x))
    elif:
        if x.__name__ in self__locals:
            # PRINT A WARNING
        self.__locals[x.__name__] = x
```

Reply:
Furthermore, if we really really wanted to, we could subdivide self.__locals:

```
self.__locals = {'modules': {}, 'anon_names': {}}
```

And then merge the two dictionaries (and do a warning printout in the case of name clashes) when performing the eval.
(I think this is more efficient than doing the check for each module that is added.

And then when describing the locals dict, you can print the names of any modules, and then all of the anonymous non-module names, to shorten the printout